pkg/alertmanager/alertmanager.go

```diff
@@ -575,3 +607,124 @@ func (t *tenantRateLimits) RateLimit() rate.Limit {
 func (t *tenantRateLimits) Burst() int {
 	return t.limits.NotificationBurstSize(t.tenant, t.integration)
 }
+
+var (
+	errTooManyAlerts = "too many alerts, limit: %d"
+	errAlertsTooBig  = "alerts too big, total size limit: %d bytes"
+)
+
+const (
+	insertFailureTooManyAlerts = "too_many_alerts"
+	insertFailureAlertsTooBig  = "alerts_too_big"
+)
+
+// alertsLimiter limits the number and size of alerts being received by the Alertmanager.
+// We consider an alert unique based on its fingerprint (a hash of its labels) and
+// its size it's determined by the sum of bytes of its labels, annotations, and generator URL.
+type alertsLimiter struct {
+	tenant string
+	limits Limits
+
+	failureCounter *prometheus.CounterVec
+
+	mx        sync.Mutex
+	sizes     map[model.Fingerprint]int
+	count     int
+	totalSize int
+}
+
+func newAlertsLimiter(tenant string, limits Limits, failureCounter *prometheus.CounterVec) *alertsLimiter {
+	return &alertsLimiter{
+		tenant:         tenant,
+		limits:         limits,
+		sizes:          map[model.Fingerprint]int{},
+		failureCounter: failureCounter,
+	}
+}
+
+func (a *alertsLimiter) PreStore(alert *types.Alert, existing bool) error {
+	if alert == nil {
+		return nil
+	}
+
+	countLimit := a.limits.AlertmanagerMaxAlertsCount(a.tenant)
+	sizeLimit := a.limits.AlertmanagerMaxAlertsSizeBytes(a.tenant)
+
+	newSize := alertSize(alert.Alert)
+
+	a.mx.Lock()
+	defer a.mx.Unlock()
+
+	// We allow existing alerts in with no checks. Alert update currently cannot change labels,
+	// annotations or generator URL. Also we want to make sure that alerts already in
+	// store can be resolved.
+	if !existing {
+		if countLimit > 0 && (a.count+1) > countLimit {
+			a.failureCounter.WithLabelValues(insertFailureTooManyAlerts).Inc()
+			return fmt.Errorf(errTooManyAlerts, countLimit)
+		}
+
+		if sizeLimit > 0 && (a.totalSize+newSize) > sizeLimit {
+			a.failureCounter.WithLabelValues(insertFailureAlertsTooBig).Inc()
+			return fmt.Errorf(errAlertsTooBig, sizeLimit)
+		}
+	}
+
+	return nil
+}
+
+func (a *alertsLimiter) PostStore(alert *types.Alert, existing bool) {
+	if alert == nil {
+		return
+	}
+
+	newSize := alertSize(alert.Alert)
+	fp := alert.Fingerprint()
+
+	a.mx.Lock()
+	defer a.mx.Unlock()
+
+	if existing {
+		a.totalSize -= a.sizes[fp]
+	} else {
+		a.count++
+	}
+	a.sizes[fp] = newSize
+	a.totalSize += newSize
+}
+
+func (a *alertsLimiter) PostDelete(alert *types.Alert) {
+	if alert == nil {
+		return
+	}
+
+	fp := alert.Fingerprint()
+
+	a.mx.Lock()
+	defer a.mx.Unlock()
+
+	a.totalSize -= a.sizes[fp]
+	delete(a.sizes, fp)
+	a.count--
+}
+
+func (a *alertsLimiter) currentStats() (count, totalSize int) {
+	a.mx.Lock()
+	defer a.mx.Unlock()
+
+	return a.count, a.totalSize
+}
+
+func alertSize(alert model.Alert) int {
```

Comment:
Have you considered `unsafe.Sizeof()` ?

Reply:
`unsafe.Sizeof` doesn’t return correct size for map, we still need to count size of keys/values manually.